content/blog/how-i-learned-to-stop-worrying-and-love-performance-regression-tests.asciidoc

```diff
@@ -0,0 +1,424 @@
+Test results are independent from wall clock time and thus provide actionable information also when not testing with production-like hardware and data volumes.
+
+This post is the first one in a series of three:
+
+* Part one provides an overview over continuous performance unit testing with JfrUnit,
+based on the examples of asserting expected memory allocation and database I/O
+* The second part will discuss the assertion of the SQL statements executed by an application,
+based on the instrumentation of JDBC APIs using JMC Agent
+* In the third and last part we'll explore how to keep historic performance test metrics,
+allowing to identify trends in these metrics over a longer period of time
+
+== Getting Started With JfrUnit
+
+JfrUnit is an extension for JUnit 5, which integrates Flight Recorder into unit tests;
+it makes it straight forward to initiate a JFR recording including a given set of event types,
+execute some test routine, and then assert the JFR events which should be produced.
+
+Here is a basic example of a JfrUnit test:
+
+[source,java]
+----
+@JfrEventTest // <1>
+public class JfrUnitTest {
+
+  public JfrEvents jfrEvents = new JfrEvents();
+
+  @Test
+  @EnableEvent("jdk.GarbageCollection") // <2>
+  @EnableEvent("jdk.ThreadSleep")
+  public void shouldHaveGcAndSleepEvents() throws Exception {
+    System.gc();
+    Thread.sleep(1000);
+
+    jfrEvents.awaitEvents(); // <3>
+
+    ExpectedEvent = event("jdk.GarbageCollection"); // <4>
+    assertThat(jfrEvents).contains(event); 
+    
+    ExpectedEvent = event("jdk.GarbageCollection") // <4>
+        .with("cause", "System.gc()"));
+    assertThat(jfrEvents).contains(event); 
+
+    ExpectedEvent = event("jdk.ThreadSleep").
+        with("time", Duration.ofSeconds(1)));
+    assertThat(jfrEvents).contains(event); 
+
+    assertThat(jfrEvents.ofType("jdk.GarbageCollection")).hasSize(1); // <5>
+  }
+}
+----
+<1> `@JfrEventTest` marks this as a JfrUnit test, activating its extension
+<2> All JFR event types to be recorded must be enabled via `@EnableEvent`
+<3> After running the test code, `awaitEvents()` must be invoked as a synchronization barrier,
+making sure all previously produced events have been received
+<4> Using the `JfrEventsAssert#event()` method, an `ExpectedEvent` instance can be created, optionally containing a set of expected attribute values, which then is asserted via `JfrEventsAssert#assertThat()`
+<5> `JfrEvents#ofType()` allows to filter on specific event types, enabling arbitrary assertions on the returned stream of ``RecordedEvent``s
+
+As we'll see in a bit, JfrUnit integrates nicely with the Java streams API,
+simplifying the filtering and aggregation of recorded events before matching them against asserted values.
+It also persists the JFR recording of each test method,
```

Comment:
"It also persists" -> does refer to what exactly here? I assume it == JfrUnit itself but that is a bit unclear IMHO :-)